libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1565,6 +1565,16 @@ void TypeChecker::endVisit(ElementaryTypeNameExpression const& _expr)
 
 void TypeChecker::endVisit(Literal const& _literal)
 {
+	if (_literal.looksLikeAddress())
+	{
+		if (_literal.passesAddressChecksum())
```

Comment:
Probably the real solution would be checking validity at the time of assignment to `address` (like we do for UTF8 to string assignment).